### PR TITLE
Support 3rd party initiated login for OpenID Connect

### PR DESCRIPTION
**Pull request by jkakavas**
https://openid.net/specs/openid-connect-core-1_0.html#ThirdPartyInitiatedLogin

### CODE REVIEW

**Comment by albertzaharovits** [comment]
This does not cover the spec you've pointed to:

> login_hint
> OPTIONAL. Hint to the Authorization Server about the login identifier the End-User might use to log in. **If the client receives a value for this string-valued parameter, it MUST include it in the Authentication Request as the login_hint parameter value.**

The transport request for the _prepare authentication request_ does not carry the `login_hint` parameter.

>target_link_uri
>OPTIONAL. URL that the RP is requested to redirect to after authentication. RPs MUST verify the value of the target_link_uri to prevent being used as an open redirector to external sites.

Do we have to keep this in the flow context and return it in the `OpenIdConnectAuthenticateResponse` ? 😒

**Comment by jkakavas** [comment]
> The transport request for the _prepare authentication request_ does not carry the `login_hint` parameter.

Yeap, thanks for keeping me honest here. I worked mostly from experience here and missed adding the login hint. It's rarely used AFAIK but it's easy to support so I'll add it. 

> Do we have to keep this in the flow context and return it in the `OpenIdConnectAuthenticateResponse` ? unamused

I don't think we should support `target_link_uri`. There aren't many use cases where the 3rd party would necessarily know a deep link within Kibana so that it would point the UA there from the beginning. Even if we do support this in the future, this should be a URI pointing to the `faciliator` only and as such, it can very well be stored in the user's session with the `facilitator`

**Comment by albertzaharovits** [comment]
> I don't think we should support `target_link_uri`

I totally agree. This is the kind of flexibility we should refrain from following.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/oidc/RestOpenIdConnectPrepareAuthenticationAction.java [hunk 1]
```diff
@@ -36,6 +36,7 @@
 
     static {
         PARSER.declareString(OpenIdConnectPrepareAuthenticationRequest::setRealmName, new ParseField("realm"));
+        PARSER.declareString(OpenIdConnectPrepareAuthenticationRequest::setIssuer, new ParseField("iss"));
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/oidc/RestOpenIdConnectPrepareAuthenticationAction.java]:
```
PARSER.declareString(OpenIdConnectPrepareAuthenticationRequest::setLoginHint, new ParseField("login_hint"));
```

**Review by albertzaharovits — APPROVED** [review]
LGTM after updating the rest-transport PARSER

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/oidc/RestOpenIdConnectPrepareAuthenticationAction.java (hunk 1 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/oidc/RestOpenIdConnectPrepareAuthenticationAction.java]: :man_facepalming:

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/oidc/OpenIdConnectPrepareAuthenticationRequest.java [hunk 2]
```diff
@@ -20,7 +20,17 @@
...
 public class OpenIdConnectPrepareAuthenticationRequest extends ActionRequest {
 
+    /**
+     * The name of the OpenID Connect realm in the configuration that should be used for authentication
+     */
     private String realmName;
+    /**
+     * In case of a
+     * <a href="https://openid.net/specs/openid-connect-core-1_0.html#ThirdPartyInitiatedLogin">3rd party initiated authentication</a>, the
+     * issuer to the UA needs to be redirected for authentication
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/oidc/OpenIdConnectPrepareAuthenticationRequest.java]:
```suggestion
     * issuer that the User Agent needs to be redirected to for authentication.
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/oidc/OpenIdConnectPrepareAuthenticationRequest.java [hunk 3]
```diff
@@ -36,10 +46,23 @@ public String getNonce() {
...
 
+    public String getIssuer() {
+        return issuer;
+    }
+
+    public String getLoginHint() {
+        return loginHint;
+    }
+
+
```
Comment (jaymode) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/oidc/OpenIdConnectPrepareAuthenticationRequest.java]:
```suggestion
```

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/oidc/TransportOpenIdConnectPrepareAuthenticationAction.java [hunk 4]
```diff
@@ -38,19 +41,39 @@ public TransportOpenIdConnectPrepareAuthenticationAction(TransportService transp
     @Override
     protected void doExecute(Task task, OpenIdConnectPrepareAuthenticationRequest request,
                              ActionListener<OpenIdConnectPrepareAuthenticationResponse> listener) {
-        final Realm realm = this.realms.realm(request.getRealmName());
-        if (null == realm || realm instanceof OpenIdConnectRealm == false) {
+        Realm realm = null;
+        if (Strings.hasText(request.getIssuer())) {
+            List<OpenIdConnectRealm> matchingRealms = this.realms.stream().filter(r -> r instanceof OpenIdConnectRealm)
```
Comment (jaymode) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/oidc/TransportOpenIdConnectPrepareAuthenticationAction.java]: can we use a single filter?

```
.filter(r -> r instanceof OpenIdConnectRealm && ((OpenIdConnectRealm) r).isIssuerValid(request.getIssuer()))
```

**Review by jaymode — APPROVED** [review]
left some minor comments, otherwise LGTM

**Pull request merged**